onnxruntime/contrib_ops/cpu/transformers/beam_search.cc

```diff
@@ -0,0 +1,650 @@
+  // Input shapes:
+  //   input_ids  : (batch_size, sequence_length)
+  //   vocab_mask : (vocab_size) or nullptr
+
+  const Tensor* input_ids = context.Input<Tensor>(0);
+  const auto& dims = input_ids->Shape().GetDims();
+  if (dims.size() != 2) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'input_ids' is expected to have 2 dimensions, got ",
+                           dims.size());
+  }
+
+  const Tensor* vocab_mask = context.Input<Tensor>(8);
+  if (vocab_mask != nullptr) {  // vocab_mask is optional
+    const auto& vocab_mask_dims = vocab_mask->Shape().GetDims();
+    if (vocab_mask_dims.size() != 1) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'vocab_mask' is expected to have 1 dimension, got ",
+                             vocab_mask_dims.size());
+    }
+
+    // There is dependency on vocab_size parameter, which shall be set before calling this function.
+    if (static_cast<int>(vocab_mask_dims[0]) != parameters_->vocab_size) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'vocab_mask' shape does not match with vocab_size, got ",
+                             vocab_mask_dims[0]);
+    }
+
+    // store vocab mask in parameters.
+    parameters_->vocab_mask = vocab_mask->DataAsSpan<int32_t>();
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+Status BeamSearchImpl<T>::Initialize() {
+  auto status = Status::OK();
+
+#define CHECK_SCALAR_INPUT(name, index, required)                                                                 \
+  auto* name##_tensor = context_.Input<Tensor>(index);                                                            \
+  if (name##_tensor) {                                                                                            \
+    if (!name##_tensor->Shape().IsScalar()) {                                                                     \
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "'BeamSearch' input " #name " should be a scalar. Got shape of ", \
+                             name##_tensor->Shape());                                                             \
+    }                                                                                                             \
+  } else if (required) {                                                                                          \
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "'BeamSearch' input " #name " is required");                        \
+  }
+
+  CHECK_SCALAR_INPUT(min_length, 1, false);
+
+  CHECK_SCALAR_INPUT(max_length, 2, true);
+
+  CHECK_SCALAR_INPUT(num_beams, 3, true);
+
+  CHECK_SCALAR_INPUT(num_return_sequences, 4, true);
+
+  CHECK_SCALAR_INPUT(temperature, 5, true);
+
+  CHECK_SCALAR_INPUT(length_penalty, 6, true);
+
+  ORT_RETURN_IF(parameters_->num_return_sequences > parameters_->num_beams, "'num_return_sequences' has to be smaller or equal to 'num_beams'.");
```

Comment:
trying to recall justification over this: every new iteration in beam search doesn't have to be extracting num_return_sequences. The search can go on with num_beams among the non-terminated sequences.

This can be done as an extention/improvement too.

Reply:
For each batch, we only keep track of num_beams results, so the number of results (per batch) to return cannot exceed num_beams.

Reply:
Beam search is only to expand the search space. Say, there is no beam search, the very first iteration can still return num_sequences with highest scored logits.

Reply:
Each iteration only generate one word, and there is no limit on token level.

The limitation is on number of sequences. I have not seen a case with num_return_sequences > num_beams.